API/files/Core/Config.php.txt
<?php
/**************************************************************************
Copyright 2017 Benato Denis

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 *****************************************************************************/

namespace Gishiki\Core;

use Gishiki\Algorithms\Strings\Manipulation;
use Gishiki\Algorithms\Collections\SerializableCollection;

/**
 * Used to parse an application file and to generate a valid configuration.
 *
 * @author Benato Denis <benato.denis96@gmail.com>
 */
final class Config
{
    /**
     * @var array|null the application configuration
     */
    protected $configuration = null;

    /**
     * @var string|null the name of the configuration file
     */
    protected $filename = null;

    /**
     * @var \Memcached|null
     */
    protected $cache = null;

    /**
     * Config constructor.
     * Parse the application configuration file when cached
     * content from a previous parse is not available.
     *
     * @param null|string     $filename the path and name of the file, or null for default
     * @param \Memcached|null $cache    the memcached caching instance. Speedup loading when provided.
     */
    public function __construct($filename = null, \Memcached $cache = null)
    {
        //set the filename
        $this->setFilename($filename);

        //set the cache
        if (!is_null($cache)) {
            $this->cache = &$cache;
        }

        //load settings
        $this->finalizeLoading();
    }

    /**
     * Load configuration faster than possible.
     *
     * If caching is enabled attempt to load configuration from there.
     * It configuration cannot be found in cache load from file and
     * update the cache.
     */
    protected function finalizeLoading()
    {
        $cacheKey = sha1($this->getFilename());

        //check if settings are available inside memcached
        if (!is_null($this->cache)) {
            $cacheContent = $this->cache->get($cacheKey);

            if ($this->cache->getResultCode() !== \Memcached::RES_NOTFOUND) {
                $this->configuration = unserialize($cacheContent);
                return;
            }
        }

        //load setting using the old-fashioned way :)
        $this->loadSettingsFromFile();

        //serialize and cache settings
        if ($this->cache instanceof \Memcached) {
            $this->cache->set($cacheKey, serialize($this->configuration), 60*60*24);
        }
    }

    /**
     * Get the path of the file containing the application configuration.
     *
     * @return string the current configuration file path
     */
    public function getFilename() : string
    {
        return $this->filename;
    }

    /**
     * Get the path that contains the application configuration file.
     *
     * @return string the settings file directory
     */
    public function getDirectory() : string
    {
        return dirname($this->getFilename());
    }

    /**
     * Set the path of the file containing the application configuration.
     *
     * @param  string|null $filename the path and name of the file, or null for default
     * @throws Exception the error preventing the file to be read
     */
    public function setFilename($filename = null)
    {
        $filename = (is_null($filename)) ? Application::getCurrentDirectory() . "settings.json" : $filename;

        if (!file_exists($filename)) {
            throw new Exception("The given configuration file cannot be read", 100);
        }

        $this->filename = $filename;
    }

    /**
     * Load the configuration from the loaded file.
     *
     * @throws Exception the error preventing the file to be read
     */
    protected function loadSettingsFromFile()
    {
        //get the json encoded application settings
        $configContent = file_get_contents($this->filename);

        //parse the settings file
        $incompleteConfig = SerializableCollection::deserialize($configContent)->all();

        //complete the request
        $this->configuration = $this->completeSettings($incompleteConfig);
    }

    /**
     * Get the application configuration.
     *
     * @return SerializableCollection the configuration
     */
    public function getConfiguration() : SerializableCollection
    {
        return new SerializableCollection($this->configuration);
    }

    /**
     * Complete the configuration resolving every value placeholder.
     *
     * Read more on documentation.
     *
     * @param  array $collection the configuration to be finished
     * @return array the completed configuration
     */
    private function completeSettings(array &$collection) : array
    {
        foreach ($collection as &$value) {
            //check for substitution
            if ((is_string($value)) && ((strpos($value, '{{@') === 0) && (strpos($value, '}}') !== false))) {
                $value = (($toReplace = Manipulation::getBetween($value, '{{@', '}}')) != '') ?
                    $this->getValueFromEnvironment($toReplace) : $value;
            } elseif (is_array($value)) {
                $value = $this->completeSettings($value);
            }
        }

        return $collection;
    }

    /**
     * Get the value of an environment variable from its name.
     *
     * @param  string $key the name of environment variable
     * @return string the value of the environment variable
     */
    protected function getValueFromEnvironment($key) : string
    {
        if (($value = getenv($key)) === false) {
            $value = constant($key);
        }

        return (string)$value;
    }
}



API/files/Core/Environment.php.txt
<?php
/**************************************************************************
Copyright 2017 Benato Denis

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 *****************************************************************************/

namespace Gishiki\Core;

/**
 * Represents the current execution environment.
 *
 * @author Benato Denis <benato.denis96@gmail.com>
 */
abstract class Environment
{
    /**
     * Check if the current execution environment
     * contains the given entry.
     *
     * @param string $key the key to be searched
     * @return bool true whether the given key is defined
     * @throws \InvalidArgumentException the given key is not valid
     */
    public static function has($key) : bool
    {
        if ((!is_string($key)) || (strlen($key) < 0)) {
            throw new \InvalidArgumentException("The given environment key must be a non-empty valid string");
        }

        return (($value = getenv($key)) !== false);
    }

    /**
     * Return the value associated with the given key
     * in the current execution environment.
     *
     * @param string $key     the key to be read
     * @param mixed  $default the default value returned when the key is not found
     * @return array|false|string the value held by the environment
     * @throws \InvalidArgumentException the given key is not valid
     */
    public static function get($key, $default = null)
    {
        if ((!is_string($key)) || (strlen($key) < 0)) {
            throw new \InvalidArgumentException("The given environment key must be a non-empty valid string");
        }

        return (static::has($key)) ? getenv($key) : $default;
    }

    /**
     * Set in the current execution environment
     * a new value with the given value: create the
     * key if it doesn't exists.
     *
     * @param string $key the key to be added
     * @param string $value the value to be associated with the key
     * @throws \InvalidArgumentException the given key is not valid
     */
    public static function set($key, $value)
    {
        if ((!is_string($key)) || (strlen($key) < 0)) {
            throw new \InvalidArgumentException("The given environment key must be a non-empty valid string");
        }

        putenv("$key=".$value);
    }

    /**
     * Remove in the current execution environment the given entry.
     *
     * @param string $key the key to be removed
     * @throws \InvalidArgumentException the given key is not valid
     */
    public static function remove($key)
    {
        if ((!is_string($key)) || (strlen($key) < 0)) {
            throw new \InvalidArgumentException("The given environment key must be a non-empty valid string");
        }

        putenv("$key");
    }
}



changelog.txt
*v2.0.0-alpha* => *v2.0.0*
- Better Code
- Fixed almost every bug preventing the framework to work (I'm sorry)
- Everything is instantiated and the entire lifecycle is contained within an Application object => I can now perform a full framework test
- A Controller has a reference to each database and logging connections
- Every unhandled exception is now logged, not only Gishiki own exceptions
- Completed the Router class

Gishiki/config.php
<?php
/**************************************************************************
Copyright 2015 Benato Denis

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*****************************************************************************/ 



/* =============================================================================
 *                                                                              *
 *                          PHP General Configuration                           *
 *                                                                              *
 * ============================================================================ */

//$Development value sets the error reporting:
//TRUE  --> Development	--> ALL PHP errors are reported
//FALSE --> Production	--> No PHP errors are reported
$Development = TRUE; //BOOLEAN

//Should the gzip compression be used when possible?
$Compression = FALSE; //BOOLEAN

//The name of the directory for the caching engine (won't be used if using APC)
$ChachingDir = "cache"; //STRING


/* =============================================================================
 *                                                                              *
 *                          Security Settings                                   *
 *                                                                              *
 * ============================================================================ */

//the default password to crypt and decrypt content (at least 32 characters)
$DefaultPassword    = "*gV8iVW5oQSIncIZOcqYjVlgxARwt]6J"; //STRING

//the name of the directory that contains the keys for the asymmetric chiper
$keyDir             = "keys"; //STRING

//the name of the file, stored inside the keys directory that contains
//the application RSA key
$RSAServerUniqueKey = "ServerKey"; //STRING


/* =============================================================================
 *                                                                              *
 *                          Database Settings                                   *
 *                                                                              *
 * ============================================================================ */

//the name of the directory that contains sqlite databases
$databaseDir            = "databases"; //STRING

//the name of the directory that contains sqlite databases
$databaseConnectionsDir = "connections"; //STRING


/* =============================================================================
 *                                                                              *
 *                          MVC Structure                                       *
 *                                                                              *
 * ============================================================================ */

//the name of the directory that contains model, view and controller (must be placed in the root)
$applicationDir         = "application"; //STRING

//the name of the directory that contains controllers
$controllerDir          = "Controller"; //STRING

//the name of the directory that contains models
$modelDir               = "Model"; //STRING

//the name of the directory that contains views
$viewDir                = "View"; //STRING

//the name of the directory that contains classes
$classDir               = "Classes"; //STRING

//the name of the directory that contains files to be served to the client
$resourcesDir           = "Resources"; //STRING

/* =============================================================================
 *                                                                              *
 *                          Routing Configuration                               *
 *                                                                              *
 * ============================================================================ */

//turn ON and OFF the routing
$RoutingEnabled = TRUE; //BOOLEAN

//simple re-routing sobstitutions
$RoutingRules = array(
    "" => "Default/Index",
    "index.php" => "Default/Index",
); //ARRAY of sobstitution: index becomes value

//regexp re-routing sobstitutions
$RoutingRegex = array(
    "-(.*)/Default-" => "{1}/Index",
    "-(.*).php-" => "Default/Index/{1}.php",
); //ARRAY matched regexp are sobstituted with the regexp number, if the number is between '{' and '}'


/* =============================================================================
 *                                                                              *
 *                          Cookies Configuration                               *
 *                                                                              *
 * ============================================================================ */

//the prefix for each cookie manageable through the framework
$CookiesPrefix = "GishikiCookie_"; //STRING

//the prefix for each cookie manageable through the framework
//this prefix is added to the cookie to distinguish normal cookies from secure cookies
$CookiesEncryptedPrefix = "84RioJRQ3IkFimmeNA"; //STRING

//the key used to encrypt cookies (at least 32 characters)
$CookiesKey = "yu2RMp7okcdM9tFWnkRWPD3laK@HPP#F"; //STRING

//the default time (in seconds) that must pass to invalidate the cookie,
//if set to 0 the cookie will be a session cookie
$DefaultCookiesExpiration = 60*60*24*30; //INTEGER

//The path on the server in which the cookie will be available on,
//if set to "/", the cookie will be available within the entire domain.
//you should edit this value to match the framework path on your server.
$CookiesPath = "/"; //STRING